developing/vcs/README.md

```diff
@@ -2,7 +2,225 @@
+on the GitHub side with [Probot](https://probot.github.io/apps/delete-merged-branch/).
+Locally, you can add an alias to your `~/.gitconfig` file. Example:
+
+  ```
+  [alias]
+    fresh = "!git co master && git pull && git branch --merged | grep -v '\\*' | xargs -n 1 git branch -d"
+  ```
+
+  Then instead of doing the typical `git checkout master && git pull`, you would
+  only run `git fresh`.
+
+### GitHub repo settings
+
+* [Protect the master branch](https://help.github.com/en/articles/configuring-protected-branches)
+by turning on the following settings at a minimum:
+
+  * Require pull request reviews before merging
+  * Require status checks to pass before merging
+  * Include administrators
+
+* Under `Settings->Merge button`:
+  * Uncheck all options except for `Allow merge commits` because it's the only
+  one that allows deleting local branches with `-d`. We should never have to use
+  `-D` to delete a branch.
+
+## PR size
+
+Pull requests should be small and focused enough to be reviewable in under 15,
+or ideally under 5, minutes.
+
+## Git workflow
+
+### Fork vs clone
+
+If it's a repo you have write and push access to, clone it. Otherwise, fork it.
+
+### Cloning with SSH vs HTTPS
+
+Do we have strong opinions on this?
+
+### Typical workflow
+
+Clone the repo locally (example below uses SSH):
+
+```
+git clone git@github.com:trussworks/Engineering-Playbook.git
+cd Engineering-Playbook
+```
+
+Create a local feature branch based off master. Start the branch name with your
+initials. For example: `mb-git-worfklow`. If the branch is related to a Pivotal
+story, add the story's ID to the end of the branch: `mb-git-workflow-166068559`
+
+```
+git checkout master
+git pull
+git checkout -b <your-initials-branch-name-optional-pivotal-story-id>
+```
+
+To rename a branch:
+
+```
+git checkout <branch-you-want-to-rename>
+git branch -m <new-branch-name>
+```
+
+Note that if there is already an active PR for your branch, and if people have
+already started commenting on it, you should not rename your branch, because
+that will create a separate branch and you will have to open a new PR for it.
+If you want to rename a branch that you already pushed to GitHub, but that
+hasn't been reviewed yet, run the following commands:
+
+```
+git checkout old_branch
+git branch -m new_branch                   # Rename branch locally
+git push origin --delete old_branch        # Delete the old remote branch
+git push --set-upstream origin new_branch  # Push the new branch, set local branch to track the new remote
+```
+
+As you work locally, before pushing to GitHub, feel free to amend and/or squash
+your commits to keep the history meaningful and concise. For example, if you
+made a commit, and then realized you made a typo, you can amend the previous
+commit:
+
+```
+git commit --amend
+```
+
+When the feature/fix is complete and tests pass, rebase to incorporate upstream
+changes.
+
+```
+git fetch origin
+git rebase origin/master
```

Comment:
A while back we had a big discussion on this flow vs merging and it never resolved. There were good points on both sides and the meeting we took a poll and it was split 50/50. I think for this document it probably makes sense to encourage merging to update your branch, and have rebase be in a separate more advanced doc, since that can lead you down some tricky paths, especially for people new to git.

Reply:
Yep! I much prefer `git merge` and am happy to talk about why!

Reply:
How do y'all squash commits when you use `merge` to pull in changes from `master`?

Reply:
Not speaking for anyone else but I don't squash pushed commits, when merging or otherwise. In general I don't think rewriting history is necessary or helpful, and the risks are significant. 

There may be distinct questions here though:

1. Is it correct to rebase a branch after code review, prior to merging (and therefore force push after rebasing)?
2. Regardless of #1, is it correct to "squash and merge" when merging to master?

I tend to think the answer to both is 'No' but my conviction about the second one is much weaker than the first. It sounds like you're asking the second question.

Reply:
Yes, I believe squashing into meaningful commits, that pass tests and other checks results in a history that's easier to read, understand, and debug. Check out my latest commit, where I add the pros and cons for both, and a list of popular repos that enforce squash and rebase: https://github.com/trussworks/Engineering-Playbook/pull/46/commits/4da0ce11c68ef2334e0a1474f59d5c0bb6b10464

Reply:
My pattern is to only squash things before I've shared the PR for review. After that, I avoid it for commits that have been shared and only use merge to update my branch. Before my branch is in PR review I rebase with master.